presto-spi/src/main/java/com/facebook/presto/spi/block/BlockFlattener.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.api.Experimental;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.spi.block.ClosingBlockLease.newLease;
+import static java.util.Objects.requireNonNull;
+
+@Experimental
+public class BlockFlattener
+{
+    private final ArrayAllocator allocator;
+
+    public BlockFlattener(ArrayAllocator allocator)
+    {
+        this.allocator = requireNonNull(allocator, "allocator is null");
+    }
+
+    /**
+     * Flattens {@code RunLengthEncodedBlock} and {@code DictionaryBlock} into the top-level type
+     * such that there is only one level of nesting.  For example, a dictionary block which consists
+     * of another dictionary block would be flattened to just one level, with all of the ids remapped
+     * accordingly.
+     */
+    public BlockLease flatten(Block block)
+    {
+        requireNonNull(block, "block is null");
+        if (block instanceof DictionaryBlock) {
+            return flattenDictionary((DictionaryBlock) block);
```

Comment:
nit: maybe call them `flattenDictionaryBlock` and `flattenRunLengthEncodedBlock`?